src/rust/ensogl/src/display/shape/compound/port.rs

```diff
@@ -0,0 +1,146 @@
+//! This module defines the shapes required for drawing node ports.
+
+use crate::data::color::*;
+use crate::display::shape::*;
+use crate::display::shape::primitive::def::modifier::immutable::*;
+use crate::display::shape::primitive::def::primitive::*;
+use crate::prelude::*;
+
+use crate::display::shape::primitive::def::class::ShapeOps;
+use crate::math::circle_segment::CircleSegment;
+use crate::math::topology::unit::AngleOps;
+use crate::math::topology::unit::Distance;
+use crate::math::topology::unit::PixelDistance;
+use crate::math::topology::unit::Pixels;
+use crate::math::topology::unit::{Angle, Degrees};
+
+use nalgebra::Vector2;
+
+
+
+// ===========================
+// === Port Specification ===
+// ===========================
+
+/// Indicates whether a port is incoming or outgoing.
+#[derive(Clone,Copy,Debug,Eq,PartialEq)]
+pub enum PortDirection {
+    /// Indicates port facing towards the center of its inner circle.
+    Inwards,
+    /// Indicates port facing away from the center of its inner circle.
+    Outwards,
+}
+
+
+/// Defines the properties of a port shape and can then
+/// be used to build the port shape.
+///
+/// Ports are constructed around an inner circle, and thus
+/// most measurements are in degrees, which are measured around
+/// a inner circle, that is defined by the `inner_radius`.
+#[derive(Clone,Copy,Debug)]
+pub struct PortSpecification{
+    /// Height of the port.
+    pub height      : f32,
+    /// Width of the port in degrees.
+    pub width       : Angle<Degrees>,
+    /// Radius of the inner circle that the port is constructed around.
+    pub inner_radius: f32,
+    /// Direction the port is facing.
+    pub direction   : PortDirection,
+    /// Location of the port along the inner circle.
+    pub location    : Angle<Degrees>,
+    /// Fill colour of the port.
+    pub color       : Srgb<f32>,
+}
+
+
+// ============
+// === Port ===
+// =============
+
+#[derive(Clone,Copy,Debug)]
+/// Dummy struct for port construction.
+pub struct Port;
+
+impl Port{
+    /// Construct a port according to the given `PortSpecification`.
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(spec:PortSpecification) -> AnyShape{
+        let show_debug = false;
+        match spec.direction{
+            PortDirection::Inwards  => Port::new_port_inwards(spec, show_debug),
+            PortDirection::Outwards => Port::new_port_outwards(spec, show_debug),
+
+        }
+    }
+
+    /// Construct an outwards facing port.
+    fn new_port_outwards(spec:PortSpecification, show_debug:bool) -> AnyShape{
```

Comment:
Why are you defining 2 shapes instead of 1 which is parametrized by the orientation (outwards / inwards)? Two separate shapes require 2 separate sprite systems, which cause 2 draw calls instead of 1 draw call. Plus, these 2 functions share a big part of very similar code.